src/crimson/common/gated.h

```diff
@@ -0,0 +1,67 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <seastar/core/gate.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/future-util.hh>
+
+#include "crimson/common/exception.h"
+#include "crimson/common/log.h"
+#include "include/ceph_assert.h"
+
+namespace crimson::common {
+
+class Gated {
+ public:
+  virtual void print(std::ostream&) const = 0;
+ protected:
+  static seastar::logger& gated_logger() {
+    return crimson::get_logger(ceph_subsys_osd);
+  }
+  template <typename Func>
+  void gated_dispatch(const char* what, Func&& func) {
+    (void) seastar::with_gate(pending_dispatch, std::forward<Func>(func)
+    ).handle_exception([this, what] (std::exception_ptr&& eptr) {
+      if (*eptr.__cxa_exception_type() == typeid(system_shutdown_exception)) {
+	gated_logger().debug(
+	    "{}, {} skipped, system shutdown", *this, what);
+	return seastar::now();
+      }
+      gated_logger().error(
+          "{} gated_dispatch() {} caught exception: {}", *this, what, eptr);
+      ceph_abort("unexpected exception from gated_dispatch()");
+    });
+  }
+  template <typename Func>
+  seastar::future<> gated_dispatch_with_future(const char* what, Func&& func) {
```

Comment:
if the only difference between these two variants is that the latter returns a future while the former discards it, why not let the one which runs in background calls the later and discard the return value?

also, i'd suggest rename the function name of the former to something like
`dispatch_in_background()`. and use `dispatch()` for the latter. as the class is named `Gated` already, there is no need to prefix the function names with "gated" anymore.